src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,640 @@
+`first_poly` and `second_poly`. Here, `first_index` and `second_index`
+are the indices of `first_poly` and `second_poly` respectively
+in the Groebner basis.
+The first and second prefix and suffix satisfy
+first_prefix g_i first_suffix = second_prefix g_j second_suffix
+where `g_i` is `first_poly` and `g_j` is `second_poly`.
+"""
+struct ObstructionTriple{T} <: Obstruction{T}
+    first_poly::FreeAssAlgElem{T}
+    second_poly::FreeAssAlgElem{T}
+    first_prefix::Monomial
+    first_suffix::Monomial
+    second_prefix::Monomial
+    second_suffix::Monomial
+    first_index::Int
+    second_index::Int
+end
+
+function ObstructionTriple{T}(first_poly::FreeAssAlgElem{T},
+                              second_poly::FreeAssAlgElem{T},
+                              pre_and_suffixes::NTuple{4, Monomial},
+                              first_index::Int,
+                              second_index::Int
+                            ) where T
+    return ObstructionTriple{T}(first_poly, second_poly, pre_and_suffixes[1], 
+                                pre_and_suffixes[2], pre_and_suffixes[3], 
+                                pre_and_suffixes[4], first_index, second_index)
+end
+
+function FreeAssAlgElem{T}(R::FreeAssAlgebra{T}, mon::Monomial) where {T}
+    return FreeAssAlgElem{T}(R, [one(base_ring(R))], [mon], 1)
+end
+
+"""
+takes an obstruction triple (p, q, o) and returns the common multiple
+of the leading terms of p and q defined by o
+TODO documentation
+"""
+function common_multiple_leading_term(ot::ObstructionTriple{T}) where {T}
+    return FreeAssAlgElem{T}(parent(ot.first_poly), ot.first_prefix) *
+           FreeAssAlgElem{T}(parent(ot.first_poly), _leading_word(ot.first_poly)) *
+           FreeAssAlgElem{T}(parent(ot.first_poly), ot.first_suffix)
+end
+
+function s_polynomial(ot::ObstructionTriple{T}) where {T}
+    first_term =
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.first_prefix) *
+        ot.first_poly *
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.first_suffix)
+    second_term =
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.second_prefix) *
+        ot.second_poly *
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.second_suffix)
+    return inv(leading_coefficient(ot.first_poly)) * first_term -
+           inv(leading_coefficient(ot.second_poly)) * second_term
+end
+
+# skip all of the extra length-checking
+function _leading_word(a::FreeAssAlgElem{T}) where {T}
+    return a.exps[1]
+end
+
+@doc """
+    gb_divides_leftmost(a::Word, aut::AhoCorasickAutomaton)
+
+If an element of the Groebner basis that is stored in `aut` divides `a`,
+return (true, a1, a2, keyword_index), where `keyword_index` is the index of the
+keyword that divides `a` such that `a = a1 aut[keyword_index] a2`.
+""" 
+function gb_divides_leftmost(a::Word, aut::AhoCorasickAutomaton)
+    match = search(aut, a)
+    if isnothing(match)
+        return (false, [], [], -1)
```

Comment:
This is not type stable. Presumably
```suggestion
        return (false, Word(), Word(), -1)
```
would be better, assuming `typeof(match.keyword_index) == Int`

Reply:
(This is not urgent, nor a merge blocker, I am just pointing out things as I see them)